packages/contracts-bedrock/contracts/L1/L2OutputOracleV2.sol

```diff
@@ -0,0 +1,331 @@
+    }
+
+    /**
+     * @notice Accepts an outputRoot and the timestamp of the corresponding L2 block.
+     *
+     * @param _outputRoot    The L2 output of the checkpoint block.
+     * @param _l2BlockNumber The L2 block number that resulted in _outputRoot.
+     * @param _l1BlockHash   A block hash which must be included in the current chain.
+     * @param _l1BlockNumber The block number with the specified block hash.
+     */
+    function proposeL2Output(
+        bytes32 _outputRoot,
+        uint256 _l2BlockNumber,
+        bytes32 _l1BlockHash,
+        uint256 _l1BlockNumber
+    ) external payable {
+        require(
+            msg.value >= MINIMUM_OUTPUT_PROPOSAL_COST,
+            "L2OutputOracleV2: minimum proposal cost not provided"
+        );
+
+        require(
+            l2Outputs[_l2BlockNumber].timestamp == 0,
+            "L2OutputOracleV2: output already proposed"
+        );
+
+        require(
+            computeL2Timestamp(_l2BlockNumber) < block.timestamp,
+            "L2OutputOracleV2: cannot propose L2 output in the future"
+        );
+
+        require(
+            _outputRoot != bytes32(0),
+            "L2OutputOracleV2: L2 output proposal cannot be the zero hash"
+        );
+
+        if (_l1BlockHash != bytes32(0)) {
+            // This check allows the proposer to propose an output based on a given L1 block,
+            // without fear that it will be reorged out.
+            // It will also revert if the blockheight provided is more than 256 blocks behind the
+            // chain tip (as the hash will return as zero). This does open the door to a griefing
+            // attack in which the proposer's submission is censored until the block is no longer
+            // retrievable, if the proposer is experiencing this attack it can simply leave out the
+            // blockhash value, and delay submission until it is confident that the L1 block is
+            // finalized.
+            require(
+                blockhash(_l1BlockNumber) == _l1BlockHash,
+                "L2OutputOracleV2: block hash does not match the hash at the expected height"
+            );
+        }
+
+        // Walk backwards in the doubly linked list
+        uint256 prevBlockNum = highestL2BlockNumber;
+        while (prevBlockNum > _l2BlockNumber) {
+            prevBlockNum = previousBlockNumber[prevBlockNum];
+        }
+        uint256 updatePrevNext = nextBlockNumber[prevBlockNum];
+        if (updatePrevNext != 0) {
+            previousBlockNumber[updatePrevNext] = _l2BlockNumber;
+        }
+        nextBlockNumber[prevBlockNum] = _l2BlockNumber;
+        nextBlockNumber[_l2BlockNumber] = updatePrevNext;
+        previousBlockNumber[_l2BlockNumber] = prevBlockNum;
+
+        // Update the highest L2 block number if necessary.
+        if (_l2BlockNumber > highestL2BlockNumber) {
+            highestL2BlockNumber = _l2BlockNumber;
+        }
+
+        bonds[_l2BlockNumber] = msg.sender;
```

Comment:
An implicit invariant worth making explicit and testing for is that this entry should be zero before writing the proposer's address over it.